brave/src/main/java/brave/baggage/BaggagePropagation.java

```diff
@@ -259,6 +240,45 @@ public FactoryBuilder add(BaggagePropagationConfig config) {
     return delegate.keys();
   }
 
+  /**
+   * Returns the key names used for propagation, including those used for the {@linkplain #keys()
+   * trace context} and {@linkplain SingleBaggageField#keyNames() baggage}. The result can be cached
+   * in the same scope as the propagation instance.
+   *
+   * <p>This is here for the remote propagation use cases:
+   * <ul>
+   *   <li>To generate constants for all key names. ex. gRPC Metadata.Key</li>
+   *   <li>To iterate fields when missing a get field by name function. ex. OpenTracing TextMap</li>
+   *   <li>To clear fields on re-usable requests. ex. JMS message</li>
+   * </ul>
+   *
+   * <h3>Details</h3>
+   * The {@code propagation} parameter is required because there may be multiple tracers with
+   * different baggage configuration. Also, {@link Propagation} instances can be wrapped, so you
+   * cannot use {@code instanceof} to identify if baggage is internally supported. For example,
+   * {@link ExtraFieldPropagation} internally wraps {@link BaggagePropagation}.
+   *
+   * <p>This is different than {@link BaggageField#getAll(TraceContext)}, as propagation keys may be
+   * different than {@link BaggageField#name() baggage field names}.
+   *
+   * @param propagation used to extract configuration
+   * @return a list of remote propagation key names used for trace context and baggage.
+   * @since 5.12
+   */
+  // On OpenTracing TextMap: https://github.com/opentracing/opentracing-java/issues/305
+  public static List<String> allKeyNames(Propagation<String> propagation) {
+    if (propagation == null) throw new NullPointerException("propagation == null");
+    // When baggage or similar is in use, the result != TraceContextOrSamplingFlags.EMPTY
+    TraceContextOrSamplingFlags emptyExtraction =
+      propagation.extractor((c, k) -> null).extract(Boolean.TRUE);
+    List<String> baggageKeyNames = ExtraBaggageContext.getAllKeyNames(emptyExtraction);
+    if (baggageKeyNames.isEmpty()) return propagation.keys();
+
+    List<String> result = new ArrayList<>(propagation.keys());
```

Comment:
It might be a tiny bit more performant to presize instead of precopy

```java
List<String> result = new ArrayList<>(propagation.keys().size()+ baggageKeyNames.size());
result.addAll(propataion.keys();
result.addAll(baggageKeyNames);
return unmodifiable(result);
```

Reply:
sure thing